### ISSUE TITLE
Allow BFS to count depth depend on Predicate

### ISSUE THREAD

**Issue by omerb4** [jgrapht/jgrapht#875]
```
 * JGraphT version: 1.2.0
 * Java version (java -version)/platform:  11.0.4
```

**Issue**

I need an option to override encounterVertex on BreadthFirstIterator and check whether I want to increment the depth depend on if the vertex is matching a predicate.

There's no option to create or modify SearchNodeData from inheritance.

**Steps to reproduce (small coding example)**

**Expected behaviour**

**Other information**

**Comment by d-michail** [comment]
From your description, I do not understand why you need to adjust the class `SearchNodeData`. You can simply override method `encounterVertex` and decide there whether to increase the depth or not. 

As a general comment, I would say that this is a very specialized request which is highly unlikely to benefit or be used by other users. I would suggest to simply copy the iterator in your code base and adjust it accordingly.

**Comment by omerb4** [comment]
I need SearchNodeData class to be accessible in order for me to override encounterVertex and putSeenData with a new depth value.

**Comment by d-michail** [comment]
@jsichi What is you opinion on this? Should we make more things protected? I guess having half of the implementation protected does not make any sense.

**Comment by omerb4** [comment]
protected to an open source project sounds legit to me.
It allows any user to customize whatever he wants instead of making ugly solution.

**Comment by jsichi** [comment]
Yeah, our use of protected is a bit of a code-smelly legacy antipattern, so it probably doesn't hurt to make it consistent :)

**Comment by omerb4** [comment]
Opened a PR: https://github.com/jgrapht/jgrapht/pull/878

**Comment by Toptachamann** [comment]
Resolved in #878.
